cpp/tests/sort/rank_test.cpp

```diff
@@ -0,0 +1,375 @@
+    fixed_width_column_wrapper<double> col3_rank  {{3, 5, 1, 3, 6, 3} , {1, 1, 1, 1, 1, 1} };
+    this->run_all_tests(rank_method::AVERAGE, asce_keep, col1_rank, col2_rank, col3_rank);
+}
+
+TYPED_TEST(Rank, average_asce_top)
+{
+    fixed_width_column_wrapper<double> col1_rank  {{4, 2, 1, 4, 6, 4} };
+    fixed_width_column_wrapper<double> col2_rank  {{4, 2, 1, 4, 6, 4} };
+    fixed_width_column_wrapper<double> col3_rank  {{3, 5, 1, 3, 6, 3} };
+    this->run_all_tests(rank_method::AVERAGE, asce_top, col1_rank, col2_rank, col3_rank);
+}
+
+TYPED_TEST(Rank, average_asce_bottom)
+{
+    fixed_width_column_wrapper<double> col1_rank  {{4, 2, 1, 4, 6, 4} };
+    fixed_width_column_wrapper<double> col2_rank  {{3, 1, 6, 3, 5, 3} };
+    fixed_width_column_wrapper<double> col3_rank  {{3, 5, 1, 3, 6, 3} };
+    this->run_all_tests(rank_method::AVERAGE, asce_bottom, col1_rank, col2_rank, col3_rank);
+}
+
+TYPED_TEST(Rank, average_desc_keep)
+{
+    fixed_width_column_wrapper<double> col1_rank  {{3, 5, 6, 3, 1, 3} };
+    fixed_width_column_wrapper<double> col2_rank  {{3, 5, -1, 3, 1, 3} , {1, 1, 0, 1, 1, 1} };
+    fixed_width_column_wrapper<double> col3_rank  {{4, 2, 6, 4, 1, 4} , {1, 1, 1, 1, 1, 1} };
+    this->run_all_tests(rank_method::AVERAGE, desc_keep, col1_rank, col2_rank, col3_rank);
+}
+
+TYPED_TEST(Rank, average_desc_top)
+{
+    fixed_width_column_wrapper<double> col1_rank  {{3, 5, 6, 3, 1, 3} };
+    fixed_width_column_wrapper<double> col2_rank  {{4, 6, 1, 4, 2, 4} };
+    fixed_width_column_wrapper<double> col3_rank  {{4, 2, 6, 4, 1, 4} };
+    this->run_all_tests(rank_method::AVERAGE, desc_top, col1_rank, col2_rank, col3_rank);
+}
+
+TYPED_TEST(Rank, average_desc_bottom)
+{
+    fixed_width_column_wrapper<double> col1_rank  {{3, 5, 6, 3, 1, 3} };
+    fixed_width_column_wrapper<double> col2_rank  {{3, 5, 6, 3, 1, 3} };
+    fixed_width_column_wrapper<double> col3_rank  {{4, 2, 6, 4, 1, 4} };
+    this->run_all_tests(rank_method::AVERAGE, desc_bottom, col1_rank, col2_rank, col3_rank);
+}
+
+TYPED_TEST(Rank, dense_asce_keep_pct)
+{
+    cudf::test::fixed_width_column_wrapper<double> col1_rank  {{0.75, 0.5, 0.25, 0.75, 1, 0.75} };
+    cudf::test::fixed_width_column_wrapper<double> col2_rank  {{2.0/3.0, 1.0/3.0, -1, 2.0/3.0, 1, 2.0/3.0} , {1, 1, 0, 1, 1, 1} };
+    cudf::test::fixed_width_column_wrapper<double> col3_rank  {{0.5, 0.75, 0.25, 0.5, 1, 0.5} , {1, 1, 1, 1, 1, 1} };
+    this->run_all_tests(rank_method::DENSE, asce_keep, col1_rank, col2_rank, col3_rank, true);
+}
+
+TYPED_TEST(Rank, dense_asce_top_pct)
+{
+    cudf::test::fixed_width_column_wrapper<double> col1_rank  {{0.75, 0.5, 0.25, 0.75, 1, 0.75} };
+    cudf::test::fixed_width_column_wrapper<double> col2_rank  {{0.75, 0.5, 0.25, 0.75, 1, 0.75} };
+    cudf::test::fixed_width_column_wrapper<double> col3_rank  {{0.5, 0.75, 0.25, 0.5, 1, 0.5} };
+    this->run_all_tests(rank_method::DENSE, asce_top, col1_rank, col2_rank, col3_rank, true);
```

Comment:
Shouldn't there be more tests to test `percentage==true` with methods other than `rank_method::DENSE`?

Reply:
All those cases are tested in pytest. no of test cases for pct is too many. 
`percentage==true` handling is different for `rank_method::DENSE`. so created this case deliberately.

Reply:
added more non-dense percentage tests.